core/src/toga/window.py

```diff
@@ -16,6 +16,11 @@
     from toga.app import App
     from toga.widgets.base import Widget
 
+    try:
+        import PIL.Image as PIL_Image
+    except ImportError:  # pragma: no cover
+        PIL_Image = None
```

Comment:
I don't think this will work - you're referencing `PIL_Image.Image` in the type.

Reply:
Hm, good point. I've never actually used static analysis on Python. Should I import the class directly, e.g. `from PIL.Image import Image as PIL_Image` (and then import the `open` function separately in files that need it)? Also, I guess I run mypy on it to check?

Reply:
Wait, unless you're conflating `PIL.Image` and `PIL.Image.Image`. The first is the module, the second is the class.

Reply:
In other words, as currently written, `PIL_Image.Image` is the correct class; the potential issue, I guess, if whether mypy would freak out when trying to check `None.Image` when Pillow isn't installed.

Reply:
pardon the interjection....but don't forget about `from __future__ import annotations`. it converts all type annotations to strings.

Reply:
At runtime, sure... but I thought that for mypy/PyCharm purposes, it needed to be a resolvable type?

Reply:
I admit ignorance here. I'll look more into it / happily accept guidance on how this works.

Reply:
Yeah, certainly...to be meaningful, it'll need to be resolved.

Pillow is in typeshed; so, I wouldn't be surprised if PyCharm "just works" because it probably just loads the entire typeshed and figures out the import whether or not Pillow is actually installed. mypy could theoretically work the same way...at least with `types-Pillow` installed.

I can give this a test tomorrow since I'm a little curious now....if someone doesn't beat me to it.

Reply:
Had a bit more time with the code.

To start, when in a `if TYPE_CHECKING` block, it is not necessary catch `ImportError`. At runtime, `TYPE_CHECKING` will always be `False` and the code will never run. During type checking, type checkers will interpret `TYPE_CHECKING` as `True` but then use tactics to deal with import errors, circular references, etc. and ultimately resolve the import.

With that in mind, we can consider the actual typing of `as_image(format)` where `format` can be a type of Toga's `Image` class of that of Pillow's. Given this, a type checker needs a way to figure out what `PIL` is.

If PyCharm and mypy cannot figure out what `PIL` is, they actually resolve the entire type definition of `format` to `Any`. In fact, in this situation, when I use `as_image(format=dict)`, neither complain (even with `mypy --strict`).

Therefore, something needs to define `PIL`. Since we're explicitly trying to avoid adding Pillow as a dependency, you can add `types-Pillow` as a dependency so IDEs and type checkers will always have access to the types for Pillow's interfaces. OTOH, mypy will tell its users to install `types-Pillow` since it knows that `PIL` is almost certainly talking about Pillow. However, IDEs will likely just be in the dark and users won't get the benefit of these types out of the box.

As for the type definition itself, i.e. `Image | PIL_Image.Image`, we're not looking for an instance of one of these classes; we're looking for the class itself. So, we should use `type[Image | PIL_Image.Image]`.

However, the full type definition of the method is still looser than what's really the behavior of the method.

The current definition:
```python
    def as_image(
        self,
        format: type[Image | PIL_Image.Image] = Image,
    ) -> Image | PIL_Image.Image:
```
This says that format can be one or the other and the output will be one or the other. But in reality, the output is directly tied to what the input is. Explicitly, the output will be an instance of the input class.

For instance, if I write something like this in my app:
```python
    def print_window(self) -> toga.Image:
        return self.main_window.as_image(format=toga.Image)
```
mypy will be upset about this:
```console
❯ mypy src/piltest/app.py --strict 
src/piltest/app.py:25: error: Incompatible return value type (got "toga.images.Image | PIL.Image.Image", expected "toga.images.Image")  [return-value]
Found 1 error in 1 file (checked 1 source file)
```
What mypy says is true....but we know the output will be a `toga.Image`. Therefore, we need to connect the input and output of this method.

To do this, we first define this image type (and this'll need to be in the `if TYPE_CHECKING` block):
```python
    ImageT = TypeVar("ImageT", bound=Image | PIL_Image.Image)
```
Then we can finally type this method:
```python
    def as_image(self, format: type[ImageT] = Image) -> ImageT: ...
```

Reply:
Another note....

This type definition:
```python
    ImageT = TypeVar("ImageT", bound=Image | PIL_Image.Image)
```
technically means that the input can be `Toga.Image` or `PIL.Image`....or any class that has `Toga.Image` or `PIL.Image` as a base class.

If you do not want users to pass subclasses, then this definition is more appropriate:

```python
    ImageT = TypeVar("ImageT", Image, PIL_Image.Image)
```

Reply:
Thanks for those notes. It's especially helpful having the detail on how PyCharm will complain, as that's the most likely way any bug reports around this will manifest :-) 

Right now, the `ImageT = TypeVar("ImageT", bound=Image | PIL_Image.Image)` definition seems fine to me - I can't think of any reason a subclass of `Image`/`PIL_Image.Image` would be a problem (or, if it is, it's the problem of the person subclassing, not *our* problem). Did you have a particular problem in mind? AFAICT, as long as `isinstance` works, and the mechanism for extracting data (e.g., `save()` continue to work, subclasses shouldn't be a problem.

However, the long game is to allow end-users to register plugins for other image types - for example, OpenCV has an [Image type](https://docs.opencv.org/3.1.0/dc/d2e/tutorial_py_image_display.html); it would be desirable for Toga to be able to use that image type without needing to explicitly accomodate every image-adjacent library. There's no common base class for an arbitrary image library, nor are we in a position to provide one; can you think of any way we would be able to make `ImageT` sufficiently dynamic that it would incorporate all user-declared types? The only thoughts I've had are:

1. Add `Any` as an option on `ImageT`. Not really helpful, but... not *untrue*.
2. Define a Protocol for a generic "Image" that is sufficiently loose that any type *could* match it, and include that in ImageT. This really isn't any better than (1), but it will have a name that will be a little more contextually relevant.
3. Try and do it dynamically - something like:
```
registered_image_classes = [ ... dynamically build a list from plugin imports... ]

if TYPE_CHECKING:
    ImageT = TypeVar("ImageT", bound=any([toga.Image] + registered_image_classes))
```
but I have no idea if that will work. At the very least, it would require (AFAICT) Toga's doc-building environment to have at least some of the plugins installed.

(To be clear, we don't need to solve this problem right now - the "fixed list" ImageT solution is fine for current purposes. But if you've got any ideas on where this can go, that would be helpful).

Reply:
> I was looking at `as_format()`; it's currently performing identity checks on the class:

Ah - good catch. They should be `issubclass`, not type checks.

>  > However, the long game is to allow end-users to register plugins for other image types
> 
> If you want to support arbitrary classes as an Image, this is begging for structural typing then. You'd just need a `Protocol` with the required minimum interface. At that point, you don't even need to import `PIL` to include it in the typing because any class that satisfies the `Protocol` would be accepted as an input.

Yeah - the problem is that I'm not sure there *is*  common protocol - every image class has its own API and shape. Even "what size is this image" is inconsistent in whether it's a `size` attribute returning a tuple, separate `width` and `height` attributes, or a `get_size()` API. And that's before you get to "how do I get the data for this image" problem, which is what this API *really* cares about.

Reply:
hmm....I was afraid that would be the case.

The ability to dynamically determine the supported image classes from plugins really starts to diverge from "status analysis" to some kind of "runtime type checking". I even wonder if the types of objects returned from loading an entry point can be inferred through the `importlib` machinery.

Nonetheless, even if the argument is typed as `Any`, at least using the `TypeVar` to connect the argument type with the return type allows user code to remain coherent with types in their code flow.

Reply:
...Dang, thanks for all the info and explanation. Will use the `ImageT = TypeVar("ImageT", bound=Image | PIL_Image.Image)` definition, pending further developments.

As to subclasses: `PIL_Image.open(buffer)` would also have to become `format.open(buffer)`, so that the returned object is of the specified class.

It also means the first branch could only return `self` when the calling and specified image classes match exactly, and it would have to create a new instance when they both are / inherit from `toga.Image` but aren't the same as each other. Even though stock Toga images are essentially immutable, I still feel like that's probably a big enough difference in behavior that maybe it should just return a new instance in every case.

Reply:
A practical consequence is that the returned image created as a "copy" of the original would presumably be created by passing the image data, so it won't have a path attribute even if the original did. I think that actually makes sense conceptually, though: you're asking to receive your image *as* something, and the new image wasn't opened from a file, it was converted from the other image.

Reply:
Okay, I believe I've applied the changes recommended for type annotations.

I had another thought about subclass support, however. How do we know the subclass's constructor hasn't changed? What if someone subclasses toga.Image and makes it require an additional argument to create? Do we just put in the docstring that you can convert to subclasses, *as long as* they accept the same arguments?

Reply:
At some point, "consenting adult" rules need to apply :-) 

If someone did what you describe, the error is going to be a pretty clear "argument X not provided", and they're going to be able to see in a stack trace that the line of code is (a) inside Toga, and (b) doesn't provide the constructor argument. Explaining the edge case is going to complicate Toga's docs, for an edge case that won't impact most people.

Reply:
That makes sense, but... crap. I hadn't even thought about the fact that PIL Images are usually made with the module-level functions like open(). Its actual constructor isn't even documented as part of the API, and looking at the source code I'm not even sure what I'd supply to it. Apparently it's [not designed to be subclassed by application code](https://mail.python.org/pipermail/image-sig/2006-March/003832.html). So... subclass conversion support for toga.Image, explicit type match for PIL?

Reply:
I think you're overthinking it :-) The success condition here is "PyCharm/mypy doesn't complain". Make a best effort in that direction, pre-emptively catch the obvious edge cases, and defer the weird stuff for if/when it surfaces as an *actual* problem for a real-world use case.

Reply:
Me? Overthink something? Never.

Reply:
Two quick things:
1) just wanted to mention adding `types-Pillow` as a dependency for Toga again; this will help ensure the typing experience is best for those using Images....but is not strictly required
2) I notice that `imageT` is being defined in several locations....that may be required because of the multi-package structure of Toga....but we may also want to consider a consolidation of these types; otherwise, this can quickly become sprawling and difficult to understand provenance of types as well as properly keeping them in sync.

Reply:
Hrm... Agreed about consolidating the ImageT definition... but I'm less convinced about `types-Pillow`. 

Pillow isn't a dependency of Toga - it's only a dev dependency. It's up to the end user whether they have Pillow installed. Including a types package when the package it's adding types for isn't necessarily installed seems... odd. 

However, if you're talking about adding `types-Pillow` as a *dev* dependency - no objection to *that* if that will make the PyCharm development experience better.

Reply:
No; I am talking about adding it as a dependency for all users.

Buried in my diatribe:
> If PyCharm and mypy cannot figure out what `PIL` is, they actually resolve the entire type definition of `format` to `Any`. In fact, in this situation, when I use as_image(format=dict), neither complain (even with mypy --strict).

So, if Toga doesn't ship with a definition of `PIL`, `ImageT` will most likely be considered the same as `Any` for most users and this is all mostly for naught.

Reply:
Had an epiphany....

If we separate the typing for Toga images vs PIL images, we can remove this issue.

For instance: 
```python
    @overload
    def as_format(self, format: type[toga.Image]) -> toga.Image: ...

    @overload
    def as_format(self, format: type[PIL.Image.Image]) -> PIL.Image.Image: ...

    def as_format(self, format: type[T]) -> T: ...
```
In this way, if the type checker doesn't know what `PIL` is, it won't affect the ability to type a `toga.Image` as input. It also means that the `as_format()` with the implementations doesn't need to define a `bound` for the `TypeVar` for `format`....and can just leave it to default to `Any`.

Reply:
I'm not wild about the prospect of including a typing package for a package that *isn't* a dependency of Toga, and doesn't provide *any* runtime protection, but *does* increase the deliverable size of Toga, just to keep PyCharm's type hints from being a little inaccurate. 

I have no problem including it as a dev dependency; beyond that, my inclination is to (a) document it as a note on the docs for Image as something that might be helpful for PyCharm users to install, and/or (b) add a `types` extra to the toga-core package so that there's an opt-in entry point to install all the typing packages (currently only Pillow, but potentially others) that might be useful (i.e., `pip install toga-core[types]` installs the typing packages that might help with Toga development)

Reply:
> If we separate the typing for Toga images vs PIL images, we can remove this issue.

+1 to this approach.

Reply:
I've converted the annotations to overloads. Based on the fact that:

- The [docs](https://docs.python.org/3/library/typing.html#typing.overload) say the implementation definition should be entirely ignored by typecheckers.
- The example code shows it with no annotations.
- mypy's [docs](https://mypy.readthedocs.io/en/stable/more_types.html#runtime-behavior) say that "a call is never type checked against the implementation."

...I've gone ahead and removed annotations entirely from the implementation defs, which means no need for the TypeVar at all.

I also added Pillow as a dependency for documentation, since sphinx was reporting errors about failed imports.

Reply:
Thanks for sticking with us; couple issues, though :/

```diff
+    TogaImageT = TypeVar("TogaImageT", bound=Image)
+
     @overload
-    def as_format(self, type: Image) -> Image:
+    def as_format(self, format: type[TogaImageT]) -> TogaImageT: ...
 
     @overload
-    def as_format(self, type: PIL.Image.Image) -> PIL.Image.Image:
+    def as_format(self, format: type[PIL.Image.Image]) -> PIL.Image.Image: ...

    def as_image(self, format=Image): ...
```

All that said, I think, at this point, we're jumping through a lot of hoops for little gain (and I was still having some edge issues with what's above). From the perspective of the API, I think the biggest benefit is user code being able to infer the type of the return value...especially given that plugins will eventually require the input to be `Any` anyway.

So, I'd call it a day with an unbound `TypeVar`. That way, user code will be able to infer the return type....which is what matters most here. The code's gonna immediately crash out if a user sends in an incompatible type.

```python
    ImageT = TypeVar("ImageT")

    def as_format(self, format: type[ImageT]) -> ImageT: ...
```

PyCharm, Pylance (via VS Code), and mypy are all able to properly infer the return type....and we don't have all these other typing shenanigans.

(note: for whatever reason, PyCharm cannot figure out a PIL image is returned when `format` is `PIL.Image.Image`....but I'm willing to chalk that up as a PyCharm bug because Pylance and mypy don't have a problem.)

Sorry about going round and round so much. I have some ideas about how to get a better handle on typing Toga's API...and hopefully guaranteeing it via CI tests. I'll take a look at it once I finally wrap up all the "running apps in CI' stuff.

Reply:
If that's the case, does the TypeVar need to be explicitly declared? The docs say to prefer declaring them via generic function syntax where doable. If nothing's bound to it, would this be equivalent?

```python
def as_format[T](self, format: type[T]) -> T: ...
```

Reply:
Unfortunately, yes; they will need to be explicit until after Python 3.11 reaches End of Life (which is expected to happen around October 2027). Otherwise, you're absolutely right; that's an equivalent typing for the method....(and thanks for the reminder that PEP 695 added this...I had forgotten all this was simplified :)

Reply:
Aha! I keep forgetting to look at multiple versions of the docs. I've gotten rid of the overloads and replaced them with the unbound TypeVar. I opted to define it in toga.images and import it for typechecking elsewhere, which... might be a little silly? Might be six of one, half dozen of the other.